client/service/src/lib.rs

```diff
@@ -332,6 +345,35 @@ async fn build_network_future<
 	}
 }
 
+/// Set storage `NODE_ALLOWLIST` means it's a permissioned network,
+/// then only connect to these well known peers.
+fn check_node_allowlist<
+	B: BlockT,
+	BE: BackendT<B>,
+	C: BlockchainEvents<B> + StorageProvider<B, BE> + HeaderBackend<B>,
+	H: sc_network::ExHashT
+> (
+	network: &sc_network::NetworkWorker<B, H>,
+	client: &Arc<C>,
+) {
+	let id = BlockId::hash(client.info().best_hash);
+	let allowlist_storage = client.storage(&id, &StorageKey(well_known_keys::NODE_ALLOWLIST.to_vec()));
+	if let Ok(Some(raw_allowlist)) = allowlist_storage {
+		let node_allowlist: Result<Vec<NodePublic>, _> = Decode::decode(&mut &raw_allowlist.0[..]);
+
+		if let Ok(node_allowlist) = node_allowlist {
+			let mut peer_ids: HashSet<PeerId> = node_allowlist.iter()
+				.filter_map(|pubkey| Ed25519PublicKey::decode(&pubkey.0).ok())
+				.map(|pubkey| PublicKey::Ed25519(pubkey).into_peer_id())
+				.collect();
+			peer_ids.remove(network.local_peer_id());
+
+			// Set only reserved peers are allowed to connect.
+			network.service().set_reserved_peers(peer_ids, true);
+		}
+	}
```

Comment:
What if the runtime later removes the key? Shouldn't we become open to everyone?

Reply:
Most of the time is in reserved only state for a permissioned network.
If they manage to remove the key by upgrading runtime, nodes need to restart to become open in current situation, it actually makes sense to me.

Reply:
> If they manage to remove the key by upgrading runtime

From the client's perspective it doesn't matter how the key is removed.

> nodes need to restart to become open in current situation, it actually makes sense to me.

In my opinion, ideally, nobody should have to restart their node under any circumstance. Once your node is started, it should be able to run forever without any restart, and if anything that prevents that from happening should be fixed.

Do you mind adding something like this?

```suggestion
	} else {
	    // Note that the situation where the storage entry previously existed but no longer
	    // does isn't handled at the moment.
	}
```

Reply:
Thanks for the inputs, it makes sense.